### PR TITLE
fix(javascript): preserve parens for type casting for sub-item

### LINKED ISSUE #4287: Add support for JSDoc based TypeScript typecasting
This is an unsupported construction for Prettier, as it rewrites the parentheses
```
const a = /** @type {myType} */ (b).map();
```

The pattern: https://github.com/Microsoft/TypeScript/wiki/JsDoc-support-in-JavaScript (the last one)

---

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACA6nDYCGUGAOuPgM4ykYwCeADnOTRBhDABZwBONDTGAK7kAllADmGfBgBSAZQAiEMH0ZTy5HjBHRiUdBgBGtKQBNTYyYQwADAAJ1GNmvivcIgqKalQT9fNwInEwiAF5w3nAAHvSBGjpQAHR6eqgAVGl6GGkYDvwYwFCCALaGPBgAPhgAgtzc+LQAvllpqHoAbgEYRaU8APLctfUmALwYALL4nIn1XhDFABQAlBgAPBgADIkArBgA-BgA2gC6GEgYAIwbGwDcHV1QcADuQw0YY+k5eWrAAcON2UwCx6ZW4AzkMG4liWiWK+HoCyi7wAfBgoks7lAQAAaEAQejaaDkZCgWJiGBYESmTjIAAcG1xMHwhkp1I4yAATLjRBIADZwACKgnYcGQADN8LzNIz6iJeZYAMLzOHIEBQaCi3GGepgADWcBgcn8YEsyEhgjguIAVuQogAhHX6w34YpwAAyYlFSAlUstIH83E03FVhmZtF50BxIE0xREZu4FtxwjgABVmcTvZLpSAxEGYAAFeriFWZ324wIAR0EIkChdcJZ92diEE0WHq9FVsSYPHampAfwgT3zAQQGf77QgVKjpmU4qzfooMDnZbxghg9DXHOXmkajSAA)

**Input:**
```jsx
// We can "cast" types to other types using a JSDoc type assertion
// by adding an `@type` tag around any parenthesized expression.

/**
 * @type {number | Array}
 */
var numberOrArray = Math.random() < 0.5 ? [] : 100;
var newArray = /** @type {array} */ (numberOrString).map(x => x);

```

**Output:**
```jsx
// We can "cast" types to other types using a JSDoc type assertion
// by adding an `@type` tag around any parenthesized expression.

/**
 * @type {number | Array}
 */
var numberOrArray = Math.random() < 0.5 ? [] : 100;
var newArray = /** @type {array} */ (numberOrString.map(x => x));

```

### PR DESCRIPTION
**Pull request by ikatyang**
Fixes #4287

### CODE REVIEW

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -5,19 +5,39 @@ const assert = require("assert");
 const util = require("../common/util");
 const comments = require("./comments");
 
-function hasClosureCompilerTypeCastComment(text, node, locEnd) {
+function hasClosureCompilerTypeCastComment(text, path, locStart, locEnd) {
```
Comment (j-f1) [inline src/language-js/needs-parens.js]: How do you feel about dropping `ClosureCompiler` from this name since it supports TS too?

Comment (ikatyang) [reply src/language-js/needs-parens.js]: Actually, that syntax seems coming from closure compiler, see Microsoft/TypeScript#5158.

Comment (m1ck2) [reply src/language-js/needs-parens.js]: Hi is this the correct place to comment.

Comment (m1ck2) [reply src/language-js/needs-parens.js]: Hi is this the correct place to comment.

Comment (thorn0) [reply src/language-js/needs-parens.js]: Salsa ≠ TypeScript (the language)

Comment (j-f1) [reply src/language-js/needs-parens.js]: What’s Salsa then?

Comment (j-f1) [reply src/language-js/needs-parens.js]: Ah, [TS support for JS](https://github.com/Microsoft/TypeScript/issues/4789).

**Pull request merged**